### ISSUE TITLE
Replying from middleware

### ISSUE THREAD

**Issue by Cinamonas** [franciscop/server#44]
I was implementing a validation middleware that replies with status code `400` and an error message when request body is invalid, but quickly saw that replying from middleware does not prevent the rest of the pipeline from executing.

Which results in an error:
```
WARNING There is an unhandled error:
ERROR Error: Can't set headers after they are sent.
    at validateHeader (_http_outgoing.js:494:11)
    at ServerResponse.setHeader (_http_outgoing.js:501:3)
    at ServerResponse.header (./node_modules/express/lib/response.js:767:10)
    at ServerResponse.send (./node_modules/express/lib/response.js:170:12)
    at stack.push.ctx (./node_modules/server/reply/reply.js:134:13)
    at Reply.exec (./node_modules/server/reply/reply.js:157:11)
    at processReturn (./node_modules/server/src/join/index.js:16:31)
    at ./node_modules/server/src/join/index.js:41:17
    at <anonymous>
```

A minimal code displaying this behavior:
```js
server(
  ctx => 'foo', // this middleware sends 'foo'
  ctx => 'bar' // but this one is executed, too, trying to send 'bar'
);
```

Is this by design?

I also tried throwing an error from middleware after setting the reply, which does bypass the rest of the middleware, but the thrown error then gets reported as unhandled in the console.

**Comment by franciscop** [comment]
Replying from middleware *should not* prevent the rest of the pipeline to execute. This is because you might want to do some clean up later on, and have to allow to process more middleware.

What are exclusive is the router. You can do this:

```js
server(
  get('/', ctx => 'foo'),
  get('/', ctx => 'bar')
);
```

And only the first one will be executed. Routes are exclusive, but middleware cannot change the execution of other middleware later on.

Now, there are two issues here IMO:

Only send the reply if the headers were not sent already. This one is tricky but definitely doable, since we want to allow this:

```js
server(
  ctx => header('a', 'b'),
  ctx => 'foo'
);
```

The second issue is about the error messages. They should be improved A LOT, and it is one of the main things that I will be working on for the 1.1.

Finally, could you share the error code please?

**Comment by Cinamonas** [comment]
@franciscop, your reasoning makes sense, but like you said – it would be great to be able not to send the reply if the headers are already set.

Regarding throwing the error, this does bypass the rest of the pipeline:
```js
server(
  ctx => {
    ctx.res.status(400).send({ error: 'invalid data provided' });
    throw 'err';
  },
  ctx => 'bar'
);
```

But then the thrown error is reported as unhandled:
```
WARNING There is an unhandled error:
ERROR err
```

So I work around this right now with my own helper:

```js
const route = handler => (ctx) => {
  if (ctx.res.statusCode) {
    return;
  }
  return handler(ctx);
};

server(
  ctx => status(400).send({ error: 'invalid data provided' }),
  route(ctx => send('bar'))
);
```

**Comment by franciscop** [comment]
TL;DR: **See the [example bug44](https://github.com/franciscop/server/tree/master/examples/bug44)** that I just added to see how I'd do what you are doing with server.

"There is an unhandled error" is part of server, I am making sure it is more clear now. About your code, I think a better way of handling these kind of situations would be to make an error handler that gets all of the failed parts.

The fact that a thrown error will skip normal middleware but can be caught later on is perfect for these situations. You can do first the validations, then the normal code and finally the `catch` situation, just as with a normal `try{} catch() {}` workflow:

```js
const server = require('server');
const { get, error } = server.router;
const { status } = server.reply;
const CustomError = require('./custom-error');

const homePageRouter = get('/',
  // Validation and checks
  ctx => {
    if (1 < 2) throw new CustomError('Some custom error', 400);
    if (2 < 3) throw new CustomError('Some other custom error', 402);
  },

  // Normal middleware here
  ctx => 'Hello world',

  // Error handling
  error(ctx => status(ctx.error.status).json({ error: ctx.error.message }))
);

server(homePageRouter);
```

Then in `custom-error.js` you would have:

```js
const CustomError = function (message, status = 500, options = {}) {
  this.message = message;
  this.status = status;
  for (let key in options) {
    this[key] = options[key];
  }
  this.code = '/custom';  // This has special meaning but it is not documented yet; just don't set it to '/server'
};
CustomError.prototype = new Error;

module.exports = CustomError;
```

Note: if all of your errors will have the code 400, you can skip the whole `custom-error` part and just do a traditional `throw new Error('Message here')`.

***

While it is still not documented and **still experimental**, the value of `this.code` inside the error can be used to catch specific kind of errors. `error('/custom', ...)` will catch those errors with a code starting by `/custom`. You could easily create `/user` for the `4xx` and `/myappname` for the `5xx` errors and handle those differently.

**Comment by franciscop** [comment]
I made the warning and error messages a lot more verbose as well, please let me know if this is a good fix for your case or if you have any other question (related or unrelated to this).
